go/pkg/rexec/rexec.go

```diff
@@ -330,17 +332,48 @@ func (ec *Context) ExecuteRemotely() {
 	ec.Metadata.RealBytesUploaded = bytesMoved
 	log.V(1).Infof("%s %s> Executing remotely...\n%s", cmdID, executionID, strings.Join(ec.cmd.Args, " "))
 	ec.Metadata.EventTimes[command.EventExecuteRemotely] = &command.TimeInterval{From: time.Now()}
-	op, err := ec.client.GrpcClient.ExecuteAndWait(ec.ctx, &repb.ExecuteRequest{
+	eg, ctx := errgroup.WithContext(ec.ctx)
+	var streamOut, streamErr sync.Once
+	op, err := ec.client.GrpcClient.ExecuteAndWaitProgress(ctx, &repb.ExecuteRequest{
 		InstanceName:    ec.client.GrpcClient.InstanceName,
 		SkipCacheLookup: !ec.opt.AcceptCached || ec.opt.DoNotCache,
 		ActionDigest:    ec.Metadata.ActionDigest.ToProto(),
+	}, func(md *repb.ExecuteOperationMetadata) {
+		if !ec.opt.StreamOutErr {
+			return
+		}
+		if name := md.GetStdoutStreamName(); name != "" {
+			streamOut.Do(func() {
+				eg.Go(func() error {
+					path := fmt.Sprintf("%s/logstreams/%s", ec.client.GrpcClient.InstanceName, name)
+					log.V(1).Infof("%s %s> Streaming to stdout from %q", cmdID, executionID, path)
+					_, err = ec.client.GrpcClient.ReadResourceTo(ctx, path, outerr.NewOutWriter(ec.oe))
+					return err
+				})
+			})
+		}
+		if name := md.GetStderrStreamName(); name != "" {
+			streamErr.Do(func() {
+				eg.Go(func() error {
+					path := fmt.Sprintf("%s/logstreams/%s", ec.client.GrpcClient.InstanceName, name)
+					log.V(1).Infof("%s %s> Streaming to stdout from %q", cmdID, executionID, path)
+					_, err = ec.client.GrpcClient.ReadResourceTo(ctx, path, outerr.NewErrWriter(ec.oe))
+					return err
+				})
+			})
+		}
 	})
 	ec.Metadata.EventTimes[command.EventExecuteRemotely].To = time.Now()
 	if err != nil {
 		ec.Result = command.NewRemoteErrorResult(err)
 		return
 	}
 
+	if err := eg.Wait(); err != nil {
+		ec.Result = command.NewRemoteErrorResult(fmt.Errorf("failure writing output streams: %v", err))
```

Comment:
Should this be a fatal error? Why not fallback to downloading the streams?

Reply:
I agree. The complexity of the solution requires a sufficient demand for it which is currently not there or we don't know about it.
Failing if either stream fails sounds like it would be similar to failing if downloading either stream fails, which is the current behaviour.